### ISSUE TITLE
How to use in existing Klein API

### ISSUE THREAD

**Issue by wsanchez** [wsanchez/txrequest#3]
We now have both request and response APIs.

Fitting this into Klein is a bit of getting a square peg into a round hole.  `IHTTPResponse`, wants to fully encapsulate a response, but Klein splits the response headers into a thing you twiddle on the request object and *separately*, the body is a thing you return:

```python
    @route(...)
    def foo(self, _request: IRequest) -> KleinRenderable:
        request = HTTPRequestFromIRequest(_request)
        response = HTTPResponseFromIRequest(_request)

        ...logic and stuff...

        # Set status and some headers
        response.setStatus(200)
        response.headers.addValue(...)

        response.setBody(...some bytes or a Fount...)

        # Convert a response to bytes or a deferred that drains into request.write
        return responseAsKleinRenderable(response)
```

Perhaps a better option is something that looks like this:

```python
    @route(...)
    def foo(self, _request: IRequest) -> KleinRenderable:
        request = HTTPRequestFromIRequest(_request)

        ...logic and stuff...

        # Build some headers
        headers = MutableHTTPHeaders()
        headers.addValue(...)

        response = HTTPResponse(
            status=200,
            headers=headers,
            body=...some bytes or a Fount...,
        )

        # Utility that shovels a response object headers into an `IRequest` and returns the body.
        return responseIntoIRequest(response, _request)
```

I like the latter because it puts all of the conversion "at the edges".  And evolving Klein into handling this natively seems doable; we can probably build a thing that wraps an `IRequest` such that it is *both* an `IRequest` and an `IHTTPRequest`, and Klein can be adapted to handle an `IHTTPResponse` as a return value.

**Comment by glyph** [comment]
`IHTTPResponse` needs to be public :).

The only reason that Klein expects the body to be returned but not a fully-formed response is that `IRequest` in Twisted just happens to be a thing you twiddle to mess around with headers.

I think that this should be a  new API; deliberately bad name so we can bikeshed later, but:

```
from klein import Klein, KleinPlusPlus

class OldAndBusted:
    router = Klein()
    @router.route("/")
    def root(self, request: IRequest) -> Union[Things Translatable To HTTP Body]:
        ...

class NewHotness:
    router = KleinPlusPlus()
    @router.route("/")
    def root(self, request: IHTTPRequest) -> Union[Things Translatable To IHTTPResponse]:
        ...
```

Alternately, `KleinPlusPlus` could be `Klein.routePlusPlus` instead, to make mixing and matching easier.

Some things which are currently translatable to HTTP bodies will still be translatable to `IHTTPResponse` but with better fidelity: for example, `twisted.web.template.Element` can just have a straightforward adapter which would include a `text/html` content type and a `200` response code by default.  If you want to tack on more stuff, you could explicitly pull in `klein.template.ElementResponse(element, status=200, headers=...)` to augment it yourself.

**Comment by moshez** [comment]
When you have a PR, I'll be happy to review!

**Comment by wsanchez** [comment]
@moshez:  Thanks.  This PR can really use a review first, since this code uses typing a lot and it would be very helpful if `mypy` tests were running in Klein when we integrate this code:

https://github.com/twisted/klein/pull/202

**Comment by wsanchez** [comment]
This issue was me wondering out loud about how to make use of `txrequest` without making change to Klein.  In part that's because making changes to `txrequest` means getting this into Klein first, and I suspect that's going to take a long time, given my experience thus far.  Maybe not so long if @moshez wants to do some reviews.

That said, I agree.  I think `routePlusPlus` is probably the right place to start, because I think `KleinPlusPlus` can be more than a just new `route` method, and I'd want to avoid doing that and then `KleinPlusPlusPlusPlus`.  :-)

Let me see if there's anything else on my to-do list (I just got the above PR passing tests again) before I start with the Klein PRs again.

**Comment by glyph** [comment]
Yeah I wish we could make Klein happen faster, but resources for that are tricky to find :(.

You could probably make a decorator that stacks underneath klein's `@route` in txrequest to start playing with it though.

**Comment by wsanchez** [comment]
Yeah… I think I can make that work.

**Comment by wsanchez** [comment]
Working on stacked route decorator in #6

**Comment by wsanchez** [comment]
OK the current state of things:

```python
    @route(...)
    @txroute
    def foo(self, request: IHTTPRequest) -> IHTTPResponse:
        request = HTTPRequestFromIRequest(_request)

        ...logic and stuff...

        # Build some headers
        headers = MutableHTTPHeaders()
        headers.addValue(("Content-Type", "text/html; charset=utf-8"))
        ...

        # Body
        body = "<html>...</html>"

        return HTTPResponse(
            status=200,
            headers=headers,
            body=body.encode("utf-8"),
        )
```

**Comment by wsanchez** [comment]
`txroute` also lets you return text, in which case it will add a `content-type: text/plain; charset=utf-8` header and encode the text as UTF-8 bytes.

That's not as useful as one might like, though, since `text/plain` isn't the most common use case.  But one downside to this API is that there's not much information to pass along with a text string object.

**Comment by wsanchez** [comment]
…and it also accept deferreds, failures, and coroutines.
